dataplane/linux/int_dataplane.go

```diff
@@ -301,28 +321,39 @@ func NewIntDataplaneDriver(config Config) *InternalDataplane {
 		config.RulesConfig.IptablesMarkNonCaliEndpoint)
 
 	// Auto-detect host MTU.
-	if mtu, err := findHostMTU(config.MTUIfacePattern); err != nil {
+	hostMTU, err := findHostMTU(config.MTUIfacePattern)
+	if err != nil {
 		log.WithError(err).Fatal("Unable to detect host MTU, shutting down")
-	} else {
-		// We found the host's MTU. Default any MTU configurations that have not been set.
-		// We default the values even if the encap is not enabled, in order to match behavior
-		// from earlier versions of Calico. However, they MTU will only be considered for allocation
-		// to pod interfaces if the encap is enabled.
-		config.hostMTU = mtu
-		if config.IPIPMTU == 0 {
-			log.Debug("Defaulting IPIP MTU based on host")
-			config.IPIPMTU = mtu - 20
-		}
-		if config.VXLANMTU == 0 {
-			log.Debug("Defaulting VXLAN MTU based on host")
-			config.VXLANMTU = mtu - 50
-		}
-		if config.Wireguard.MTU == 0 {
-			log.Debug("Defaulting Wireguard MTU based on host")
-			config.Wireguard.MTU = mtu - 60
+	}
+	// We found the host's MTU. Default any MTU configurations that have not been set.
+	// We default the values even if the encap is not enabled, in order to match behavior
+	// from earlier versions of Calico. However, they MTU will only be considered for allocation
+	// to pod interfaces if the encap is enabled.
+	config.hostMTU = hostMTU
+	if config.IPIPMTU == 0 {
+		log.Debug("Defaulting IPIP MTU based on host")
+		config.IPIPMTU = hostMTU - ipipMTUOverhead
+	}
+	if config.VXLANMTU == 0 {
+		log.Debug("Defaulting VXLAN MTU based on host")
+		config.VXLANMTU = hostMTU - vxlanMTUOverhead
+	}
+	if config.Wireguard.MTU == 0 {
+		// The default MTU on Azure is 1500, but the underlying network stack will fragment packets at 1400 bytes, see
+		// https://docs.microsoft.com/en-us/azure/virtual-network/virtual-network-tcpip-performance-tuning#azure-and-vm-mtu
+		// for details.
+		// Additionally, Wireguard sets the DF bit on its packets, and so if the MTU is set too high large packets will
+		// be dropped.
+		// Therefore it is necessary to allow for the difference between the MTU of the host and the underlying network.
+		if config.KubernetesProvider == ProviderAKS && config.RouteSource == "WorkloadIPs" {
+			log.Debug("Defaulting Wireguard MTU based on host and AKS with WorkloadIPs")
+			config.Wireguard.MTU = hostMTU - aksMTUOverhead - wireguardMTUOverhead
 		}
+		log.Debug("Defaulting Wireguard MTU based on host")
+		config.Wireguard.MTU = hostMTU - wireguardMTUOverhead
```

Comment:
This code will still run and clobber the MTU even if you go through the above conditional

Reply:
Even with the `return nil` I just added on line 327?

Reply:
Good catch!